File: mbfpp/code/ScalaExamples/src/main/scala/com/mblinn/mbfpp/functional/memoization/Examples.scala
package com.mblinn.mbfpp.functional.memoization

object Examples {

  def expensiveLookup(id: Int) = {
    Thread.sleep(1000)
    println(s"Doing expensive lookup for $id")
    Map(42 -> "foo",  12 -> "bar",  1 -> "baz").get(id)
  }
  
  def memoizeExpensiveLookup() = {
    var cache = Map[Int, Option[String]]()
    (id: Int) => 
      cache.get(id) match {
        case Some(result: Option[String]) => result
        case None => {
          val result = expensiveLookup(id)
          cache += id -> result
          result
        }
      }
  }
  val memoizedExpensiveLookup = memoizeExpensiveLookup
}

File: mbfpp/code/ClojureExamples/src/mbfpp/functional/memoization/examples.clj
(ns mbfpp.functional.memoization.examples)

(defn expensive-lookup [id]
  (Thread/sleep 1000)
  (println (str "Lookup for " id))
  ({42 "foo" 12 "bar" 1 "baz"} id))

(def memoized-expensive-lookup 
  (memoize expensive-lookup))

(def slow-fib
  (fn [n]
    (cond
      (<= n 0) 0
      (< n 2) 1
      :else (+ (slow-fib (- n 1)) (slow-fib (- n 2))))))

(def mem-fib
  (memoize
    (fn [n]
      (cond
        (<= n 0) 0
        (< n 2) 1
        :else (+ (mem-fib (- n 1)) (mem-fib (- n 2)))))))
